addons/mass_mailing/models/mailing.py

```diff
@@ -1325,16 +1330,41 @@ def _convert_inline_images_to_urls(self, body_html):
 
         if did_modify_body:
             return lxml.html.tostring(root, encoding='unicode')
-        return body_html
+        return html_content
 
     def _create_attachments_from_inline_images(self, b64images):
         if not b64images:
             return []
 
-        attachments = self.env['ir.attachment'].create([{
-            'datas': b64image,
-            'name': f"cropped_image_mailing_{self.id}_{i}",
-            'type': 'binary',} for i, b64image in enumerate(b64images)])
+        ir_attachment = self.env['ir.attachment']
+        existing_attachments = dict(ir_attachment.search([
+            ('res_model', '=', 'mailing.mailing'),
+            ('res_id', '=', self.id),
+        ]).mapped(lambda record: (record.checksum, record)))
+
+        attachments, vals_for_attachs = [], []
+        next_img_id = len(existing_attachments)
+        for b64image in b64images:
+            checksum = ir_attachment._compute_checksum(base64.b64decode(b64image))
+            existing_attach = existing_attachments.get(checksum)
+            # Existing_attach can be None, in which case it acts as placeholder
+            # for attachment to be created.
+            attachments.append(existing_attach)
+            if not existing_attach:
+                vals_for_attachs.append({
+                    'datas': b64image,
+                    'name': f"image_mailing_{self.id}_{next_img_id}",
+                    'type': 'binary',
+                    'res_id': self.id,
+                    'res_model': 'mailing.mailing'
+                })
+                next_img_id += 1
+
+        # Stack with new attachments.
+        new_attachments = list(ir_attachment.create(vals_for_attachs))[::-1]
```

Comment:
Instead of this whole thing, could we maybe use `create_unique`? It seems to serve this purpose already